### PR TITLE
gh-2157 Operations implement equals()

### PR DESCRIPTION
**Pull request by c015dariu**
#2157 

Added equals and hashcode to all operations tested by OperationTest.

Fixed test code for some special cases where new Class() would result in null exceptions for certain fields.

Ran checkstyle fixes, these have picked up extra cases in other files however/

### CODE REVIEW

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/join/Join.java [hunk 1]
```diff
@@ -135,6 +137,44 @@ public void setCollectionLimit(final Integer collectionLimit) {
...
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        final Join<?> join = (Join<?>) o;
+
+        return new EqualsBuilder()
+                .append(leftSideInput, join.leftSideInput)
+                .append(rightSideOperation, join.rightSideOperation)
+                .append(matchMethod, join.matchMethod)
```
Comment (d47853) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/join/Join.java]: Same thing

https://github.com/gchq/Gaffer/blob/master/core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/join/match/ElementMatch.java

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/io/MultiInput.java [hunk 2]
```diff
@@ -67,7 +67,7 @@ default B input(final Iterable<? extends I_ITEM> input) {
             if (null != _getOp().getInput()) {
                 throw new IllegalStateException("Input has already been set");
             }
-            _getOp().setInput((Iterable) input);
```
Comment (d47853) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/io/MultiInput.java]: Why is this being done as part of this issue?

File: core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/GetVariableTest.java [hunk 3]
```diff
@@ -37,7 +36,7 @@ public void shouldShallowCloneOperation() {
 
         final GetVariable opClone = op.shallowClone();
 
-        assertNotEquals(op, opClone);
```
Comment (d47853) [inline core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/GetVariableTest.java]: This test is trying to test that the clone is not the same object as the original object. The test should be updated to reflect this

File: core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/JoinTest.java [hunk 4]
```diff
@@ -75,7 +74,7 @@ public void shouldShallowCloneOperation() {
         // When
         final Join clone = op.shallowClone();
 
-        assertNotEquals(op, clone);
+        assertEquals(op, clone);
```
Comment (d47853) [inline core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/JoinTest.java]: Same for all of these

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/add/AddSchemaToLibrary.java [hunk 5]
```diff
@@ -94,6 +96,36 @@ public void setParentSchemaIds(final List<String> parentSchemaIds) {
...
+        }
+
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        final AddSchemaToLibrary that = (AddSchemaToLibrary) o;
+
+        return new EqualsBuilder()
+                .append(schema, that.schema)
```
Comment (d47853) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/add/AddSchemaToLibrary.java]: Schema class does not implement equals

File: library/hdfs-library/src/main/java/uk/gov/gchq/gaffer/hdfs/operation/AddElementsFromHdfs.java [hunk 6]
```diff
@@ -273,6 +275,59 @@ public AddElementsFromHdfs shallowClone() {
...
+            return false;
+        }
+
+        final AddElementsFromHdfs other = (AddElementsFromHdfs) obj;
+
+        return new EqualsBuilder()
+                .append(failurePath, other.failurePath)
+                .append(workingPath, other.workingPath)
+                .append(validate, other.validate)
+                .append(inputMapperPairs, other.outputPath)
```
Comment (d47853) [inline library/hdfs-library/src/main/java/uk/gov/gchq/gaffer/hdfs/operation/AddElementsFromHdfs.java]: This line doesn't look correct

File: library/sketches-library/src/test/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/DoublesUnionSerialiserTest.java [hunk 7]
```diff
@@ -37,7 +37,7 @@
     @SuppressWarnings("unchecked")
```
Comment (d47853) [inline library/sketches-library/src/test/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/DoublesUnionSerialiserTest.java]: All of the whitespace changes should be removed

File: library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/graphframe/GetGraphFrameOfElements.java [hunk 8]
```diff
@@ -124,6 +126,36 @@ public GetGraphFrameOfElements shallowClone() {
...
+        }
+
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        final GetGraphFrameOfElements that = (GetGraphFrameOfElements) o;
+
+        return new EqualsBuilder()
+                .append(converters, that.converters)
```
Comment (d47853) [inline library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/graphframe/GetGraphFrameOfElements.java]: Converters don't implement equals

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/Operation.java [hunk 9]
```diff
@@ -190,7 +190,7 @@ default void close() throws IOException {
     default ValidationResult validate() {
         final ValidationResult result = new ValidationResult();
 
-        HashSet<Field> fields = Sets.<Field>newHashSet();
```
[the remark this answers is not in the document]
Comment (c015dariu) [reply core/operation/src/main/java/uk/gov/gchq/gaffer/operation/Operation.java]: Not sure how that got in there, will remove.

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java [hunk 10]
```diff
@@ -77,15 +79,45 @@ public void setInput(final I input) {
...
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        final Map<?, ?> map = (Map<?, ?>) o;
+
+        return new EqualsBuilder()
+                .append(input, map.input)
+                .append(options, map.options)
+                .append(functions, map.functions)
```
[the remark this answers is not in the document]
Comment (c015dariu) [reply core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Map.java]: To use the Serialiser I'll need to surround with a try/catch. In the event of a failed serialisation, should I simply return false or ignore functions  in equals or is there a logger/error report I should perform in that block?

**Pull request closed without merging**